src/Microsoft.SqlTools.ServiceLayer/Agent/Notebooks/Exec-Notebook.ps1

```diff
@@ -0,0 +1,211 @@
+$DatabaseQueryHashTable["OutputAs"] = "DataTables"
+
+$CellExcecutionCount = 1
+
+foreach ($NotebookCell in $TemplateNotebookJsonObject.cells) {
+    $NotebookCellOutputs = @()
+    <#
+    Ignoring Markdown or raw cells
+    #>
+    if ($NotebookCell.cell_type -eq "markdown" -or $NotebookCell.cell_type -eq "raw" -or $NotebookCell.source -eq "") {
+        continue;
+    }
+    <#
+    Getting the source T-SQL from the cell
+    #>
+    $DatabaseQueryHashTable["Query"] = $NotebookCell.source
+    <#
+    Executing the T-SQL Query and storing the result and the time taken to execute
+    #>
+    $SqlQueryExecutionTime = Measure-Command { $SqlQueryResult = @(Invoke-Sqlcmd @DatabaseQueryHashTable  4>&1) }
+    <#
+    Setting the the Notebook Cell Execution Count
+    #>
+    $NotebookCell.execution_count = $CellExcecutionCount++
+    $NotebookCellTableOutputs = @()
+    if ($SqlQueryResult) {
+        foreach ($SQLQueryResultElement in $SqlQueryResult) {
+            <#
+            Iterating over the results by Invoke-Sqlcmd
+            There are 2 types of errors:
+                1. Verbose Output: Print Statements:
+                These needs to be added to the beginning of the cell outputs
+                2. Datatables from the database
+                These needs to be added to the end of cell outputs
+            #>
+            switch ($SQLQueryResultElement.getType()) {
+                System.Management.Automation.VerboseRecord {
+                    <#
+                    Adding the print statments to the cell outputs
+                    #>
+                    $NotebookCellOutputs += $(ParseStringToNotebookOutput($SQLQueryResultElement.Message))
+                }
+                System.Data.DataTable {
+                    <#
+                    Storing the print Tables into an array to be added later to the cell outpuyt
+                    #>
+                    $NotebookCellTableOutputs += $(ParseTableToNotebookOutput $SQLQueryResultElement  $CellExcecutionCount)
+                }
+                <#
+                TODO Throw an appropriate error
+                #>
+                Default { }
+            }
+        }
+    }
+    if ($SqlQueryError) {
+        <#
+        Adding the parsed query error from Invoke-Sqlcmd
+        #>
+        $NotebookCellOutputs += $(ParseQueryErrorToNotebookOutput($SqlQueryError))
+    }
+    if ($SqlQueryExecutionTime) {
+        <#
+        Adding the parsed execution time from Measure-Command 
+        #>
+        $NotebookCellExcutionTimeString = "Total execution time: " + $SqlQueryExecutionTime.ToString("hh\:mm\:ss\.fff")
+        $NotebookCellOutputs += $(ParseStringToNotebookOutput($NotebookCellExcutionTimeString))
+    }
+    <#
+    Adding the data tables
+    #>
+    $NotebookCellOutputs += $NotebookCellTableOutputs
+    $NotebookCell.outputs = $NotebookCellOutputs
+    
+}
+
+
+return ($TemplateNotebookJsonObject | ConvertTo-Json -Depth 100)
```

Comment:
Why the limit? This should be documented at the very least if you think it's necessary.

Reply:
I'm hoping it's fairly short since most of the outputs should be strings, correct?

Reply:
Yeah, I think the maximum depth of  nested JSON objects in jupyter notebooks is 5 or 6. I'll check that and make changes accordingly

Reply:
I'd just remove the limit if it's only going to be 5 or 6, that way we don't have to update that if it ever changes later. And with the depth being that small it doesn't seem like enough of a concern to warrant explicitly limiting.

Reply:
The default limit of ConvertTo-Json cmdlet is 2, which is why I had to overwrite it.